core/server/master/src/main/java/alluxio/master/file/meta/LockedInodePath.java

```diff
@@ -213,69 +310,235 @@ public synchronized void downgradeLastWithScheme(LockingScheme lockingScheme) {
+    newComponents[components.length] = component;
+    return newComponents;
+  }
+
+  /**
+   * Returns a copy of the path with the final edge write locked. This requires that we haven't
+   * already locked the final edge, i.e. the path is incomplete.
+   *
+   * @return the new locked path
+   */
+  public LockedInodePath lockFinalEdgeWrite() throws InvalidPathException {
+    Preconditions.checkState(!fullPathExists());
+
+    LockedInodePath newPath =
+        new LockedInodePath(mUri, mInodes, mLockList, mPathComponents, LockPattern.WRITE_EDGE);
+    newPath.traverse();
+    return newPath;
+  }
+
+  private void traverseOrClose() throws InvalidPathException {
+    try {
+      traverse();
+    } catch (Throwable t) {
+      close();
+      throw t;
     }
-    return lockedDescendantPath;
   }
 
   /**
-   * Unlocks the last inode that was locked.
+   * Traverses the inode path according to its lock pattern. If the inode path is already partially
+   * traversed, this method will pick up where the previous traversal left off. If the path already
+   * ends in a write lock, traverse will populate the inodes list without taking any additional
+   * locks.
+   *
+   * On return, all existing inodes in the path are added to mInodes and the inodes are locked
+   * according to to {@link LockPattern}.
    */
-  public synchronized void unlockLast() {
-    if (mLockList.isEmpty()) {
+  public void traverse() throws InvalidPathException {
+    if (mLockList.getLockMode() == LockMode.WRITE) {
+      traverseWithoutLocking();
       return;
     }
-    mLockList.unlockLast();
+
+    // This locks the root edge and inode if necessary.
+    bootstrapTraversal();
+
+    // Each iteration either locks a new inode/edge or hits a missing inode and returns.
+    while (!fullPathExists()) {
+      int lastInodeIndex = mLockList.getInodes().size() - 1;
+      String nextComponent = mPathComponents[lastInodeIndex + 1];
+      boolean isFinalComponent = lastInodeIndex == mPathComponents.length - 2;
+
+      if (mLockList.endsInInode()) { // Lock an edge next.
+        if (mLockPattern == LockPattern.WRITE_EDGE && isFinalComponent) {
+          mLockList.lockEdge(nextComponent, LockMode.WRITE);
+        } else {
+          mLockList.lockEdge(nextComponent, LockMode.READ);
+        }
+      } else { // Lock an inode next.
+        InodeView lastInode = mLockList.getInodes().get(lastInodeIndex);
+        if (!lastInode.isDirectory()) {
+          throw new InvalidPathException("Traversal failed. Component " + lastInodeIndex + "("
+              + lastInode.getName() + ") is a file");
+        }
+        InodeView nextInode = lastInode.asDirectory().getChild(nextComponent);
+        if (nextInode == null && mLockPattern == LockPattern.WRITE_EDGE && !isFinalComponent) {
+          // This pattern requires that we obtain a write lock on the final edge, so we must
+          // upgrade to a write lock.
+          mLockList.unlockLastEdge();
```

Comment:
Read locks cannot be atomically upgraded to write locks because if two readers try to perform the upgrade at the same time, they will immediately deadlock. We need to do the upgrade non-atomically, then handle the possibility that the world has changed from under us.

Reply:
the lines of code following the write lock are dedicated to handling the case where something happened after the write lock